internal/config/variables/variables.go

```diff
@@ -0,0 +1,531 @@
+					Subject:  attr.Expr.Range().Ptr(),
+				})
+				val = cty.DynamicVal
+			}
+		}
+
+		v.Default = &InputValue{
+			Source: sourceDefault,
+			Value:  val,
+		}
+
+		// It's possible no type attribute was assigned so lets make sure we
+		// have a valid type otherwise there could be issues parsing the value.
+		if v.Type == cty.NilType {
+			v.Type = val.Type()
+		}
+	}
+
+	return &v, diags
+}
+
+// SetJobInputValues collects values set via the CLI (-var, -varfile) and
+// local env vars (WP_VAR_*) and translates those values to pb.Variables. These
+// pb.Variables can then be set on the job for eventual parsing on the runner,
+// after the runner has decoded the variables defined in the waypoint.hcl.
+// All values are set as protobuf strings, with the expectation that later
+// evaluation will convert them to their defined types.
+func SetJobInputValues(vars map[string]string, files []string) ([]*pb.Variable, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	ret := []*pb.Variable{}
+
+	// The order here is important, as the order in which values are evaluated
+	// dictate their precedence. We therefore evalute these three sources in order
+	// of env, file, cli.
+
+	// process env values ("env" source)
+	{
+		env := os.Environ()
+		for _, raw := range env {
+			if !strings.HasPrefix(raw, varEnvPrefix) {
+				continue
+			}
+			raw = raw[len(varEnvPrefix):] // trim the prefix
+
+			eq := strings.Index(raw, "=")
+			if eq == -1 {
+				// Seems invalid, so we'll ignore it.
+				continue
+			}
+
+			name := raw[:eq]
+			rawVal := raw[eq+1:]
+
+			ret = append(ret, &pb.Variable{
+				Name:   name,
+				Value:  &pb.Variable_Str{Str: rawVal},
+				Source: &pb.Variable_Env{},
+			})
+		}
+	}
+
+	// process -var-file args ("file" source)
+	for _, file := range files {
+		if file != "" {
+			pbv, diags := parseFileValues(file, sourceFile)
+			if diags.HasErrors() {
+				return nil, diags
+			}
+			ret = append(ret, pbv...)
+		}
+	}
+
+	// process -var args ("cli" source)
+	for name, val := range vars {
+		ret = append(ret, &pb.Variable{
+			Name:   name,
+			Value:  &pb.Variable_Str{Str: val},
+			Source: &pb.Variable_Cli{},
+		})
+	}
+	return ret, diags
+}
+
+// EvalInputValues evaluates the provided variable values and validates their
+// types per the type declared in the waypoint.hcl for that variable name.
+// The order in which values are evaluated corresponds to their precedence, with
+// higher precedence values overwriting lower precedence values.
+func EvalInputValues(
+	pbvars []*pb.Variable,
+	vs map[string]*Variable,
+) (InputValues, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	iv := InputValues{}
+
+	for v, def := range vs {
+		iv[v] = def.Default
+	}
+
+	for _, pbv := range pbvars {
+		variable, found := vs[pbv.Name]
+		if !found {
+			// TODO krantzinator: this should be a DiagWarning rather than a DiagErr
+			// What to do with a warning diag type?
```

Comment:
I think error is fine. It was annoying that this became an error in TF because so much workflow was already built on it being a warning, but since we haven't shipped anything yet, might as well start super strict.

Reply:
Cool cool, TODO removed